app/src/main/java/de/danoeh/antennapod/dialog/FilterDialog.java

```diff
@@ -21,42 +29,143 @@ public FilterDialog(Context context, FeedItemFilter feedItemFilter) {
+                if (item.values[0].filterId.equals(nextItem)) {
+                    item.values[0].setSelected(true);
+                    item.values[1].setSelected(false);
+                    textView2.setSelected(true);
+                    textView1.setSelected(false);
+                    textView2.setTextColor(Color.WHITE);
+                    textView1.setTextColor(Color.DKGRAY);
+                    textView3.setBackgroundResource(R.drawable.x_custom_on);
+                }
             }
+
+            textView1.setOnClickListener(arg0 -> {
+                item.values[1].setSelected(true);
+                item.values[0].setSelected(false);
+                textView2.setSelected(false);
+                textView2.setTextColor(Color.DKGRAY);
+                textView1.setSelected(true);
+                textView1.setTextColor(Color.WHITE);
+                textView3.setBackgroundResource(R.drawable.x_custom_on);
+            });
+            textView2.setOnClickListener(arg0 -> {
+                item.values[0].setSelected(true);
+                item.values[1].setSelected(false);
+                textView1.setSelected(false);
+                textView1.setTextColor(Color.DKGRAY);
+                textView2.setSelected(true);
+                textView2.setTextColor(Color.WHITE);
+                textView3.setBackgroundResource(R.drawable.x_custom_on);
+            });
+            textView3.setOnClickListener(arg0 -> {
+                item.values[0].setSelected(false);
+                item.values[1].setSelected(false);
+                textView2.setTextColor(Color.DKGRAY);
+                textView2.setSelected(false);
+                textView1.setTextColor(Color.DKGRAY);
+                textView1.setSelected(false);
+                textView3.setBackgroundResource(R.drawable.x_custom_off);
+            });
+            tableLayout.addView(row);
         }
 
-        AlertDialog.Builder builder = new AlertDialog.Builder(context);
-        builder.setTitle(R.string.filter);
-        builder.setMultiChoiceItems(items, checkedItems, (dialog, which, isChecked) -> {
-            if (isChecked) {
-                filterValues.add(values[which]);
-            } else {
-                filterValues.remove(values[which]);
-            }
-        });
+
         builder.setPositiveButton(R.string.confirm_label, (dialog, which) -> {
+            filterValues.clear();
+            for (FeedItemEnum item : FeedItemEnum.values()) {
+                for (int i = 0; i < item.values.length; i++) {
+                    if (item.values[i].getSelected()) {
+                        Log.d("**added***", "" + item.values[i].filterId);
+                        filterValues.add(item.values[i].filterId);
+                    }
+                }
+            }
             updateFilter(filterValues);
         });
+
         builder.setNegativeButton(R.string.cancel_label, null);
         builder.create().show();
     }
 
+    public enum FeedItemEnum {
```

Comment:
I would make the filters independent from the filter UI. This allows (among other things) to automatically test filters instead of having to start the UI every time. Please move this to a new class `de.danoeh.antennapod.core.feed.FeedItemFilterGroup`. Also, those enums should not store data. If you directly add/remove the `filterId` in the `filterValues` array, you do not need to store state in the enums themselves. Storing state there can lead to problems when re-opening the dialog for different podcasts.

Reply:
okay